File: valhalla/readers.py
"""
Log file readers for plain text and HTML Discord DM exports.
"""

import re
import html
from dataclasses import dataclass, field
from typing import List, Tuple, Optional
from datetime import datetime


@dataclass
class ParsedMessage:
    """Structured result from reader.read(), replacing bare 3-tuple."""
    timestamp: str
    clean_text: str
    bot_tx_signatures: List[str]
    target_tx_signatures: List[str] = field(default_factory=list)
    target_wallet_address: Optional[str] = None


class PlainTextReader:
    """Parse plain text Discord DM logs with [HH:MM] Author: format"""

    # Pattern to split messages: [HH:MM] or [YYYY-MM-DDTHH:MM] at start of line
    MESSAGE_SPLIT = re.compile(r'^(?=\[(?:\d{4}-\d{2}-\d{2}T)?\d{2}:\d{2}\])', flags=re.MULTILINE)
    # Pattern to extract author from first line (supports both timestamp formats)
    AUTHOR_PATTERN = re.compile(r'^\[((?:\d{4}-\d{2}-\d{2}T)?\d{2}:\d{2})\]\s*(.+?):\s*\n', flags=re.MULTILINE)
    # Solscan TX signature from [https://solscan.io/tx/SIG] (fallback, captures all)
    SOLSCAN_TX_PATTERN = re.compile(r'\[https://solscan\.io/tx/([A-Za-z0-9]+)\]')
    # Labeled Solscan URLs: captures label and signature
    # Matches "Target Tx 1 [https://solscan.io/tx/SIG]" and "Your Solscan 1 [...]"
    # Uses [\w \t]+? to avoid matching across newlines
    LABELED_SOLSCAN_PATTERN = re.compile(
        r'([\w \t]+?)\s*\[https://solscan\.io/tx/([A-Za-z0-9]+)\]'
    )
    # lpagent.io portfolio URL: captures wallet address
    LPAGENT_PATTERN = re.compile(
        r'\[https://app\.lpagent\.io/portfolio\?address=([A-Za-z0-9]+)\]'
    )
    # Any URL in square brackets (for stripping)
    URL_BRACKET_PATTERN = re.compile(r'\[https?://[^\]]+\]')

    def __init__(self, file_path: str):
        self.file_path = file_path
        self.header_date: Optional[str] = None

    def _parse_messages_from_text(self, text: str, date_prefix: str) -> List[ParsedMessage]:
        """
        Parse a plain-text Discord log into ParsedMessage objects.

        Args:
            text: Plain text content (date header already stripped).
            date_prefix: ISO date string from the file header (e.g. "2024-01-15"),
                         used to prefix bare HH:MM timestamps.  Pass '' if none.

        Returns:
            List of ParsedMessage for every Valhalla-authored message found.
        """
        raw_messages = self.MESSAGE_SPLIT.split(text)
        results = []

        for raw_msg in raw_messages:
            if not raw_msg.strip():
                continue

            # Extract author from first line
            author_match = self.AUTHOR_PATTERN.match(raw_msg)
            if not author_match:
                continue

            timestamp_str = author_match.group(1)  # "15:08"
            author = author_match.group(2)  # "APL. Valhalla Bot"

            # Filter: only Valhalla messages
            if 'valhalla' not in author.lower():
                continue

            timestamp = f"[{timestamp_str}]"

            # Extract target wallet address before stripping URLs
            lpagent_match = self.LPAGENT_PATTERN.search(raw_msg)
            target_wallet_address = lpagent_match.group(1) if lpagent_match else None

            # Extract labeled tx signatures before stripping URLs
            target_tx_signatures = []
            bot_tx_signatures = []
            for label, sig in self.LABELED_SOLSCAN_PATTERN.findall(raw_msg):
                label_lower = label.lower().strip()
                if 'target' in label_lower:
                    target_tx_signatures.append(sig)
                else:
                    bot_tx_signatures.append(sig)

            # Fallback: only when labeled extraction found neither bot nor target
            # signatures (prevents double-counting URLs that were already captured)
            if not bot_tx_signatures and not target_tx_signatures:
                bot_tx_signatures = self.SOLSCAN_TX_PATTERN.findall(raw_msg)

            # Strip URLs in brackets and the author line prefix
            text_body = raw_msg[author_match.end():]  # Remove the [HH:MM] Author: line
            clean_text = self.URL_BRACKET_PATTERN.sub('', text_body)

            results.append(ParsedMessage(
                timestamp=timestamp,
                clean_text=clean_text,
                bot_tx_signatures=bot_tx_signatures,
                target_tx_signatures=target_tx_signatures,
                target_wallet_address=target_wallet_address
            ))

        return results

    def read(self) -> List[ParsedMessage]:
        """Returns list of ParsedMessage objects (timestamp, clean_text, signatures, etc.)"""
        with open(self.file_path, 'r', encoding='utf-8-sig') as f:
            content = f.read()

        # Check for date header (YYYYMMDD) at the top of the file
        lines = content.split('\n')
        date_prefix = ''
        if lines and lines[0].strip():
            first_line = lines[0].strip()
            # Match 8 digits (YYYYMMDD)
            if re.match(r'^\d{8}$', first_line):
                # Validate it's a real date
                try:
                    year = int(first_line[0:4])
                    month = int(first_line[4:6])
                    day = int(first_line[6:8])
                    datetime(year, month, day)
                    # Valid date found
                    self.header_date = f"{year:04d}-{month:02d}-{day:02d}"
                    date_prefix = self.header_date
                    # Strip the date line from content
                    content = '\n'.join(lines[1:])
                except ValueError:
                    # Not a valid date, continue with full content
                    pass

        return self._parse_messages_from_text(content, date_prefix)


class HtmlReader(PlainTextReader):
    """Parse HTML Discord DM logs (from browser clipboard) with [HH:MM] Author: format"""

    def html_to_text(self, raw_html: str) -> str:
        """Convert HTML to clean text, extracting links as TEXT [URL]"""
        content = raw_html

        # Extract CF_HTML fragment if present
        start = content.find('<!--StartFragment-->')
        end = content.find('<!--EndFragment-->')
        if start >= 0 and end > start:
            content = content[start + len('<!--StartFragment-->'):end]

        # Replace links: <a href="URL">TEXT</a> → TEXT [URL]
        def replace_link(m):
            url = m.group(1)
            text = re.sub(r'<[^>]+>', '', m.group(2))  # strip tags inside anchor
            text = re.sub(r'\s+', ' ', text).strip()
            if text:
                return f'{text} [{url}]'
            else:
                return f'[{url}]'

        content = re.sub(
            r'<a\b[^>]*href\s*=\s*["\']([^"\']+)["\'][^>]*>(.*?)</a>',
            replace_link, content, flags=re.IGNORECASE | re.DOTALL
        )

        # HTML decode
        content = html.unescape(content)

        # Block elements → newlines
        content = re.sub(r'(?i)<\s*br\s*/?\s*>', '\n', content)
        content = re.sub(r'(?i)</\s*(div|p|li|tr|h[1-6])\s*>', '\n', content)

        # Strip remaining tags
        content = re.sub(r'<[^>]+>', '', content)

        # Clean whitespace
        content = re.sub(r'\n[ \t]+', '\n', content)
        content = re.sub(r'[ \t]{2,}', ' ', content)
        content = re.sub(r'\n{3,}', '\n\n', content)

        return content.strip()

    def read(self) -> List[ParsedMessage]:
        """Convert HTML to text, then use PlainTextReader logic"""
        with open(self.file_path, 'r', encoding='utf-8-sig', errors='ignore') as f:
            raw_html = f.read()

        # Check for date header before HTML content
        lines = raw_html.split('\n')
        date_prefix = ''
        if lines and lines[0].strip():
            first_line = lines[0].strip()
            # Match 8 digits (YYYYMMDD) before any HTML tags
            if re.match(r'^\d{8}$', first_line):
                # Validate it's a real date
                try:
                    year = int(first_line[0:4])
                    month = int(first_line[4:6])
                    day = int(first_line[6:8])
                    datetime(year, month, day)
                    # Valid date found
                    self.header_date = f"{year:04d}-{month:02d}-{day:02d}"
                    date_prefix = self.header_date
                    # Strip the date line from content
                    raw_html = '\n'.join(lines[1:])
                except ValueError:
                    # Not a valid date, continue with full content
                    pass

        # Convert HTML to plain text, then delegate to shared parsing logic
        plain_text = self.html_to_text(raw_html)
        return self._parse_messages_from_text(plain_text, date_prefix)


def detect_input_format(file_path: str) -> str:
    """Auto-detect if file is HTML or plain text"""
    with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
        first_4k = f.read(4096)

    # Check for HTML markers
    if re.search(r'<html|<!DOCTYPE|<!--StartFragment|<div|<span|<a\s+href', first_4k, re.IGNORECASE):
        return 'html'
    return 'text'

File: web/src/lib/readers.ts
/**
 * Log readers for plain text and HTML Discord DM exports.
 * Browser-only: takes string content, no file I/O.
 */

export interface ParsedMessage {
  timestamp: string;     // "[HH:MM]" or "[YYYY-MM-DDTHH:MM]"
  text: string;          // cleaned message body
  txSignatures: string[];
}

// Regex patterns
const MESSAGE_SPLIT = /^(?=\[(?:\d{4}-\d{2}-\d{2}T)?\d{2}:\d{2}\])/gm;
const AUTHOR_PATTERN = /^\[((?:\d{4}-\d{2}-\d{2}T)?\d{2}:\d{2})\]\s*(.+?):\s*\n/m;
const URL_BRACKET_PATTERN = /\[https?:\/\/[^\]]+\]/g;

/**
 * Convert HTML to clean plain text, extracting links as TEXT [URL].
 * Handles clipboard HTML format with StartFragment/EndFragment markers.
 */
export function htmlToText(rawHtml: string): string {
  let content = rawHtml;

  // Extract CF_HTML fragment if present
  const startIdx = content.indexOf('<!--StartFragment-->');
  const endIdx = content.indexOf('<!--EndFragment-->');
  if (startIdx >= 0 && endIdx > startIdx) {
    content = content.slice(startIdx + '<!--StartFragment-->'.length, endIdx);
  }

  // Replace links: <a href="URL">TEXT</a> -> TEXT [URL]
  content = content.replace(
    /<a\b[^>]*href\s*=\s*["']([^"']+)["'][^>]*>(.*?)<\/a>/gi,
    (_, url, inner) => {
      const text = inner.replace(/<[^>]+>/g, '').replace(/\s+/g, ' ').trim();
      return text ? `${text} [${url}]` : `[${url}]`;
    }
  );

  // Convert <time datetime="2026-02-14T20:02:53.534Z">21:02</time> to [YYYY-MM-DDTHH:MM]
  // Must happen BEFORE stripping HTML tags. Converts UTC to local timezone.
  content = content.replace(
    /<time[^>]*datetime="([^"]+)"[^>]*>[\s\S]*?<\/time>/g,
    (_, utcStr) => {
      try {
        const dt = new Date(utcStr);
        if (isNaN(dt.getTime())) return '';
        const y = dt.getFullYear();
        const mo = String(dt.getMonth() + 1).padStart(2, '0');
        const d = String(dt.getDate()).padStart(2, '0');
        const h = String(dt.getHours()).padStart(2, '0');
        const m = String(dt.getMinutes()).padStart(2, '0');
        return `[${y}-${mo}-${d}T${h}:${m}]`;
      } catch {
        return '';
      }
    }
  );

  // Fallback: if no full datetime timestamps were found, try Discord snowflake IDs
  // Snowflake format in HTML: chat-messages-CHANNELID-MESSAGEID or data-list-item-id="chat-messages__SNOWFLAKE"
  if (!content.includes('T') || !/\[\d{4}-\d{2}-\d{2}T\d{2}:\d{2}\]/.test(content)) {
    const snowflakeMatch = content.match(/chat-messages-\d+-(\d{17,20})/);
    if (snowflakeMatch) {
      const snowflakeId = BigInt(snowflakeMatch[1]);
      const discordEpoch = BigInt(1420070400000);
      const unixMs = Number((snowflakeId >> BigInt(22)) + discordEpoch);
      const dt = new Date(unixMs);
      const y = dt.getFullYear();
      const mo = String(dt.getMonth() + 1).padStart(2, '0');
      const d = String(dt.getDate()).padStart(2, '0');
      const baseDate = `${y}-${mo}-${d}`;
      // Inject full datetime into time-only markers [HH:MM] -> [YYYY-MM-DDTHH:MM]
      content = content.replace(/\[(\d{2}:\d{2})\]/g, `[${baseDate}T$1]`);
    }
  }

  // HTML decode using browser's built-in decoder
  const textarea = typeof document !== 'undefined' ? document.createElement('textarea') : null;
  if (textarea) {
    textarea.innerHTML = content;
    content = textarea.value;
  } else {
    // Fallback for SSR/testing
    content = content
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'")
      .replace(/&nbsp;/g, ' ');
  }

  // Block elements -> newlines
  content = content.replace(/<\s*br\s*\/?>/gi, '\n');
  content = content.replace(/<\/\s*(div|p|li|tr|h[1-6])\s*>/gi, '\n');

  // Strip remaining HTML tags
  content = content.replace(/<[^>]+>/g, '');

  // Clean whitespace
  content = content.replace(/\n[ \t]+/g, '\n');
  content = content.replace(/[ \t]{2,}/g, ' ');
  content = content.replace(/\n{3,}/g, '\n\n');

  return content.trim();
}

/**
 * Detect if content is HTML or plain text.
 */
export function detectFormat(content: string): 'html' | 'text' {
  const first4k = content.slice(0, 4096);
  if (/<html|<!DOCTYPE|<!--StartFragment|<div|<span|<a\s+href/i.test(first4k)) {
    return 'html';
  }
  return 'text';
}

/**
 * Check for YYYYMMDD date header at the start of content.
 * Returns { date, content } where date is "YYYY-MM-DD" or null.
 */
function extractDateHeader(content: string): { date: string | null; content: string } {
  const lines = content.split('\n');
  if (lines.length > 0 && lines[0].trim()) {
    const firstLine = lines[0].trim();
    if (/^\d{8}$/.test(firstLine)) {
      const year = parseInt(firstLine.slice(0, 4));
      const month = parseInt(firstLine.slice(4, 6));
      const day = parseInt(firstLine.slice(6, 8));
      // Validate date
      const testDate = new Date(year, month - 1, day);
      if (testDate.getFullYear() === year && testDate.getMonth() === month - 1 && testDate.getDate() === day) {
        return {
          date: `${year.toString().padStart(4, '0')}-${month.toString().padStart(2, '0')}-${day.toString().padStart(2, '0')}`,
          content: lines.slice(1).join('\n')
        };
      }
    }
  }
  return { date: null, content };
}

/**
 * Parse messages from text content (plain text format).
 * Filters to only Valhalla bot messages.
 * Returns array of ParsedMessage and optional header date.
 */
export function parseMessages(rawContent: string): { messages: ParsedMessage[]; headerDate: string | null } {
  // Check for date header
  const { date: headerDate, content } = extractDateHeader(rawContent);

  // Split into raw messages
  const rawMessages = content.split(MESSAGE_SPLIT);
  const results: ParsedMessage[] = [];

  for (const rawMsg of rawMessages) {
    if (!rawMsg.trim()) continue;

    // Extract author from first line
    const authorMatch = rawMsg.match(AUTHOR_PATTERN);
    if (!authorMatch) continue;

    const timestampStr = authorMatch[1]; // "15:08" or "2026-02-12T15:08"
    const author = authorMatch[2];

    // Filter: only Valhalla messages
    if (!author.toLowerCase().includes('valhalla')) continue;

    const timestamp = `[${timestampStr}]`;

    // Extract Solscan signatures before stripping URLs
    const txSignatures: string[] = [];
    let txMatch;
    const txRegex = /\[https:\/\/solscan\.io\/tx\/([A-Za-z0-9]+)\]/g;
    while ((txMatch = txRegex.exec(rawMsg)) !== null) {
      txSignatures.push(txMatch[1]);
    }

    // Strip URLs in brackets and the author line prefix
    const text = rawMsg.slice(authorMatch[0].length);
    const cleanText = text.replace(URL_BRACKET_PATTERN, '');

    results.push({ timestamp, text: cleanText, txSignatures });
  }

  return { messages: results, headerDate };
}

/**
 * Parse content that may be HTML or plain text.
 * Auto-detects format and processes accordingly.
 */
export function parseContent(rawContent: string): { messages: ParsedMessage[]; headerDate: string | null } {
  const format = detectFormat(rawContent);

  if (format === 'html') {
    // Check for date header before HTML conversion
    const { date: headerDate, content: htmlContent } = extractDateHeader(rawContent);
    const plainText = htmlToText(htmlContent);
    const result = parseMessages(plainText);
    // Use the header date from the original HTML if found
    return {
      messages: result.messages,
      headerDate: headerDate || result.headerDate
    };
  }

  return parseMessages(rawContent);
}
